### PR TITLE
capnp-c++: implement asGeneric method for outer template contexts

### PR DESCRIPTION
**Pull request by katreniak**
https://github.com/sandstorm-io/capnproto/issues/334

### CODE REVIEW

**Comment by katreniak** [comment]
This change solves my problem in #334. Bootstraps regenerated, tests passed locally.

I struggled a bit how to add the functionality to capnp-c++ and to keep the style. Please, let me know if you have better proposal.

File: c++/src/capnp/compiler/capnpc-c++.c++ [hunk 1]
```diff
@@ -1719,6 +1724,56 @@ private:
...
+              "    return castAs<", innerType.size() ? "typename " : "", returnType(), ">();\n"
+              "  }\n"
+              "\n");
+          break;
+      }
+    }
+    KJ_IF_MAYBE(p, templateContext.getParent()) {
+      up = makeAsGenericDef(role, *p, p->getName(), kj::strTree(
+          templateContext.hasParams() ? "::template " : "::", type,
+          templateContext.args()).flatten());
```
Comment (kentonv) [inline c++/src/capnp/compiler/capnpc-c++.c++]: Seems like the `.flatten()` is not needed here?

Comment (kentonv) [inline c++/src/capnp/compiler/capnpc-c++.c++]: Oh I see, this is the argument to makeAsGenericDef(). Perhaps the `innerType` parameter should be type `StringTree`?

Comment (kentonv) [inline c++/src/capnp/compiler/capnpc-c++.c++]: Eh, this is an irrelevant nitpick. Never mind.

**Comment by kentonv** [comment]
Looks great! Thanks.

File: c++/src/capnp/compiler/capnpc-c++.c++ (hunk 1 above)
Comment (katreniak) [inline c++/src/capnp/compiler/capnpc-c++.c++]: The code consumes innerType two times. That did not work with StringTree move only semantics. String and flatten() did the job ...

**Pull request merged**